validator/client/validator_test.go

```diff
@@ -969,3 +970,233 @@ func TestService_ReceiveBlocks_SetHighest(t *testing.T) {
 	v.ReceiveBlocks(ctx, connectionErrorChannel)
 	require.Equal(t, slot, v.highestValidSlot)
 }
+
+func TestValidator_CheckDoppelGanger(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+	flgs := featureconfig.Get()
+	flgs.EnableDoppelGanger = true
+	reset := featureconfig.InitWithReset(flgs)
+	defer reset()
+	tests := []struct {
+		name            string
+		validatorSetter func(t *testing.T) *validator
+		err             string
+	}{
+		{
+			name: "no doppelganger",
+			validatorSetter: func(t *testing.T) *validator {
+				client := mock.NewMockBeaconNodeValidatorClient(ctrl)
+				km := genMockKeymanger(10)
+				keys, err := km.FetchValidatingPublicKeys(context.Background())
+				assert.NoError(t, err)
+				db := dbTest.SetupDB(t, keys)
+				resp := &ethpb.DoppelGangerRequest{ValidatorRequests: []*ethpb.DoppelGangerRequest_ValidatorRequest{}}
+				for _, k := range keys {
+					pkey := k
+					att := createAttestation(10, 12)
+					rt, err := att.Data.HashTreeRoot()
+					assert.NoError(t, err)
+					assert.NoError(t, db.SaveAttestationForPubKey(context.Background(), pkey, rt, att))
+					resp.ValidatorRequests = append(resp.ValidatorRequests, &ethpb.DoppelGangerRequest_ValidatorRequest{PublicKey: pkey[:], Epoch: att.Data.Target.Epoch, SignedRoot: rt[:]})
+				}
+				v := &validator{
+					validatorClient: client,
+					keyManager:      km,
+					db:              db,
+				}
+				client.EXPECT().CheckDoppelGanger(
+					gomock.Any(), // ctx
+					gomock.Any(), // request
```

Comment:
Could you add the actual requests for all test cases? This will add a bit more rigidity to these tests

Reply:
So this is actually not trivial, I tried to do it previously, however the mock key manager actually uses a map underneath to store all the keys. When you request for all validating pubkeys, the ordering is non-deterministic, therefore any request mocked in here will be non deterministic too and basically only succeed 1/10 times. If needed I can create a new deterministic key manager to mock this.

Reply:
Added a custom matcher that will help with this.